streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java

```diff
@@ -867,7 +873,25 @@ private void initializeAndRestorePhase() {
         log.debug("Idempotent restore call done. Thread state has not changed.");
     }
 
+    // Check if the topology has been updated since we last checked, ie via #addNamedTopology or #removeNamedTopology
+    private void checkForTopologyUpdates() {
+        if (lastSeenTopologyVersion < topologyMetadata.topologyVersion()) {
+            lastSeenTopologyVersion = topologyMetadata.topologyVersion();
+            taskManager.handleTopologyUpdates();
+
+            log.info("StreamThread has detected an update to the topology, triggering a rebalance to refresh the assignment");
```

Comment:
very nit: just add a TODO that we can improve this case to not always enforce rebalance when version bumped, in case we forgot in future PRs?